src/freezethaw.rs

```diff
@@ -0,0 +1,58 @@
+use linux_raw_sys::ioctl::FIFREEZE;
+use linux_raw_sys::ioctl::FITHAW;
+
+use rustix::fd::AsFd;
+use rustix::{io, ioctl};
+
+fn ioctl_fifreeze<Fd: AsFd>(fd: Fd) -> io::Result<()> {
+    // SAFETY: `FIFREEZE` is a no-argument opcode.
+    unsafe {
+        let ctl = ioctl::NoArg::<{ FIFREEZE }>::new();
+        ioctl::ioctl(fd, ctl)
+    }
+}
+
+fn ioctl_fithaw<Fd: AsFd>(fd: Fd) -> io::Result<()> {
+    // SAFETY: `FITHAW` is a no-argument opcode.
+    unsafe {
+        let ctl = ioctl::NoArg::<{ FITHAW }>::new();
+        ioctl::ioctl(fd, ctl)
+    }
+}
+
+/// syncfs() doesn't flush the journal on XFS,
+/// and since GRUB2 can't read the XFS journal, the system
+/// could fail to boot.
+///
+/// http://marc.info/?l=linux-fsdevel&m=149520244919284&w=2
+/// https://github.com/ostreedev/ostree/pull/1049
+///
+/// When calling this function you should be protected by SignalTerminationGuard
```

Comment:
We could accept such a guard as an argument to this function, or honestly pretty safely just create another one?